# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [2]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [3]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [4]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [5]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [6]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [7]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [8]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [9]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [10]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [11]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [12]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [13]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [14]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [15]:
# Write your code here
df.nunique()

Season     32
Daynum    133
Wteam     364
Wscore    130
Lteam     364
Lscore    117
Wloc        3
Numot       7
dtype: int64

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [16]:
# Write your code here
df['Wteam'].value_counts()

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [17]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [18]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [19]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [20]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [21]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [22]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [23]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [24]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [25]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [26]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [27]:
# Write your code here
dfs = pd.DataFrame(df)
dfw = pd.DataFrame (df)
dfws = pd.DataFrame(df)
dfs=dfs.sort_values('Season')
dfw=dfw.sort_values('Wteam')
dfws=dfws.sort_values('Wscore')
print(dfws)
dfs.iloc[100:200,0]
dfw.iloc[100:200,2]
dfws.iloc[100:200,3]
dfs=dfs.iloc[100:200]
dfw=dfw.iloc[100:200]
dfws=dfws.iloc[100:200]
print(dfws)




        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336     2001     126   1206      35   1423      33    N      0
5364      1986      77   1229      35   1166      34    A      0
20246     1990      47   1374      36   1426      31    H      0
118590    2011     130   1336      36   1458      33    N      0
...        ...     ...    ...     ...    ...     ...  ...    ...
20022     1990      40   1116     166   1109     101    H      0
24341     1991      47   1328     172   1258     112    H      0
19653     1990      30   1328     173   1109     101    H      0
17867     1989      92   1258     181   1109     150    H      0
24970     1991      68   1258     186   1109     140    H      0

[145289 rows x 8 columns]
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
99013     2008      33   1460      43   1139      42    H      0
126840    2013      77   1366      43   1126      40    H      

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [28]:
# Write your code here
print(dfws.iloc[[dfws['Wscore'].argmax()]])
print(dfw.iloc[[dfw['Wscore'].argmax()]])
print(dfs.iloc[[dfs['Wscore'].argmax()]])


       Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
90572    2006      82   1416      45   1409      44    H      0
       Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
26437    1991     103   1102     103   1311      94    H      0
      Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
2366    1985      96   1260     105   1139      90    H      0


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [29]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [30]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [31]:
# Write your code here
df['ScoreDifference']= (df['Wscore']- df['Lscore'])
SD_mean =df.loc[:, 'ScoreDifference'].mean()
nydf=df[(df['ScoreDifference'] > SD_mean)]
print(len(df))
print(len(nydf))


145289
57227


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [32]:
# Write your code here

print(nydf.iloc[[nydf['ScoreDifference'].argmax()]])

       Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
44653    1996      48   1409     141   1341      50    H      0   

       ScoreDifference  
44653               91  


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [33]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [34]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [35]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [36]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [37]:
# Write your code here
df.groupby('Season')['Wscore'].mean()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [38]:
# Write your code here
df.groupby('Wteam')['Wscore'].max()


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [39]:
# Write your code here
df.groupby(['Season','Wteam'])['Wscore'].max()

Season  Wteam
1985    1102      85
        1103      88
        1104     100
        1106      90
        1108     112
                ... 
2016    1460      87
        1461      84
        1462      98
        1463      99
        1464     103
Name: Wscore, Length: 10172, dtype: int64

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [40]:
# Write your code here
df.groupby(['Season','Wteam'])['Wteam'].count()

Season  Wteam
1985    1102      5
        1103      9
        1104     21
        1106     10
        1108     19
                 ..
2016    1460     20
        1461     12
        1462     27
        1463     21
        1464      9
Name: Wteam, Length: 10172, dtype: int64

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [57]:
# Write your code here
#df.groupby(['Season','Wteam'])['Wteam'].count()
season_team_wins = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# DataFrame by the 'Season' column and retrieves the index of the team with the maximum number of wins for each season.
max_wins = season_team_wins.groupby('Season')['Wins'].idxmax()

# The resulting DataFrame, will contain the team with the most wins for each season.
teams_with_most_wins = season_team_wins.loc[max_wins, ['Season', 'Wteam', 'Wins']]

teams_with_most_wins.head(7)


Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [42]:
# Write your code here
ScoreDiff = [df.groupby('Wteam')['Wscore'].mean()- df.groupby('Lteam')['Lscore'].mean()]
#Q= ScoreDiff > 0
#([df.groupby('Lteam')['Lscore'].mean(),df.groupby('Wteam')['Wscore'].mean()]),df['ScoreDiff']
import numpy
arr = numpy.array(ScoreDiff)
arr = arr[arr<0]
print(ScoreDiff)
print(arr)




[Wteam
1101    17.524904
1102    10.691697
1103    11.722421
1104    11.451627
1105    13.285521
          ...    
1460    12.448554
1461    10.508110
1462    11.689903
1463    10.890888
1464     8.738993
Length: 364, dtype: float64]
[]


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [43]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [44]:
# Write your code here
df['HighScoringGame'] = 'No'  # Setting initial value of column to 'No'

for index, row in df.iterrows():
    if row['Wscore'] > 100:
        df.at[index,'HighScoringGame'] = 'Yes'
        
    if index < 15:
       print(df.at[index,'HighScoringGame'])
        
    



No
No
No
No
No
No
No
No
No
No
Yes
No
No
No
No


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [45]:
# Write your code here

games_played = {}  # Dictionary to store the number of games played by each team

for index, row in df.iterrows():
    # Increment the count for winning team
    if row['Wteam'] in games_played:
        games_played[row['Wteam']] += 1
    else:
        games_played[row['Wteam']] = 1

    # Increment the count for losing team
    if row['Lteam'] in games_played:
        games_played[row['Lteam']] += 1
    else:
        games_played[row['Lteam']] = 1        
   

# Print the total number of games played by each team
for team, games in games_played.items():
    print(f"Team {team}: Total games played = {games}")

Team 1228: Total games played = 992
Team 1328: Total games played = 968
Team 1106: Total games played = 855
Team 1354: Total games played = 906
Team 1112: Total games played = 981
Team 1223: Total games played = 363
Team 1165: Total games played = 833
Team 1432: Total games played = 69
Team 1192: Total games played = 908
Team 1447: Total games played = 903
Team 1218: Total games played = 931
Team 1337: Total games played = 922
Team 1226: Total games played = 847
Team 1242: Total games played = 993
Team 1268: Total games played = 969
Team 1260: Total games played = 914
Team 1133: Total games played = 949
Team 1305: Total games played = 922
Team 1424: Total games played = 974
Team 1307: Total games played = 969
Team 1288: Total games played = 925
Team 1344: Total games played = 951
Team 1438: Total games played = 952
Team 1374: Total games played = 916
Team 1411: Total games played = 903
Team 1412: Total games played = 962
Team 1397: Total games played = 963
Team 1417: Total games played

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [46]:
# Write your code here
highest_score_diff_per_season = {}  # Dictionary to store the highest score difference for each season

for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']

    if season not in highest_score_diff_per_season:
        highest_score_diff_per_season[season] = {
            'GameID': index,
            'ScoreDiff': score_diff,
            'WinningTeam': row['Wteam'],
            'LosingTeam': row['Lteam']
        }
    else:
        if score_diff > highest_score_diff_per_season[season]['ScoreDiff']:
            highest_score_diff_per_season[season] = {
                'GameID': index,
                'ScoreDiff': score_diff,
                'WinningTeam': row['Wteam'],
                'LosingTeam': row['Lteam']
            }

# Print the game with the highest score difference for each season
for season, game_info in highest_score_diff_per_season.items():
    print(f"Season {season}: GameID {game_info['GameID']}, Highest Score Difference = {game_info['ScoreDiff']}")
   #  print(f"Winning Team: {game_info['WinningTeam']}, Losing Team: {game_info['LosingTeam']}")
    

Season 1985: GameID 236, Highest Score Difference = 60
Season 1986: GameID 4731, Highest Score Difference = 84
Season 1987: GameID 8240, Highest Score Difference = 73
Season 1988: GameID 12046, Highest Score Difference = 68
Season 1989: GameID 16677, Highest Score Difference = 70
Season 1990: GameID 19502, Highest Score Difference = 76
Season 1991: GameID 25161, Highest Score Difference = 68
Season 1992: GameID 27997, Highest Score Difference = 82
Season 1993: GameID 33858, Highest Score Difference = 81
Season 1994: GameID 36404, Highest Score Difference = 69
Season 1995: GameID 39858, Highest Score Difference = 74
Season 1996: GameID 44653, Highest Score Difference = 91
Season 1997: GameID 49033, Highest Score Difference = 80
Season 1998: GameID 52600, Highest Score Difference = 66
Season 1999: GameID 57618, Highest Score Difference = 75
Season 2000: GameID 60616, Highest Score Difference = 75
Season 2001: GameID 65999, Highest Score Difference = 75
Season 2002: GameID 70737, Highest 

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [47]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [48]:
# Write your code here
df['TotalNumberOfGamesPerTeam'] = df.groupby('Wteam')['Lteam'].transform('count')

print(df.loc[df['Wteam'] == 1242])

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
7         1985      25   1242      58   1268      56    N      0   
30        1985      26   1242      66   1332      49    N      0   
207       1985      33   1242      86   1178      64    H      0   
519       1985      42   1242      81   1354      54    H      0   
636       1985      47   1242      87   1222      75    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
144790    2016     119   1242      86   1400      56    A      0   
144996    2016     124   1242      85   1235      78    H      0   
145174    2016     129   1242      85   1243      63    N      0   
145227    2016     130   1242      70   1124      66    N      0   
145267    2016     131   1242      81   1452      71    N      0   

        ScoreDifference HighScoringGame  TotalNumberOfGamesPerTeam  
7                     2              No                        804  
30                   17              No      

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [49]:
# Write your code here
df.loc[df.groupby('Season')['ScoreDifference'].idxmax()]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410      87    H      0   
57618     1999      58   1268     132   1317      57    H      0   
60616     2000      19   1261     112   1212      37    H      0   
65999     2001      41   1196     125   1197      50    H      0   
70737     2002      50   1328     109   1183      50    H      0   
74791     2003      33   1403     107   1311      35    H      0   
79181     2004      29   1328      94   1115      24    H      0   
83372     2005      18   1403     119   1421      55    H      0   
89542     2006      60   1400     110   1341      38    H      0   
94507     2007      59   1401     101   1212      27    H      0   
97737     2008       4   1182     129   1224      59    H      0   
103723    2009      27   1314     116   1421      48    H      0   
108405    2010      15   1397     124   1421      49    H      0   
114015    2011      22   1388     121   1152      52    H      0   
118699    2012      11   1345      96   1296      34    H      0   
124057    2013       7   1235      98   1105      40    H      0   
131373    2014      61   1256     126   1255      52    H      0   
134590    2015      11   1181     113   1342      44    H      0   
140040    2016      12   1139     144   1154      71    H      0   

        ScoreDifference HighScoringGame  TotalNumberOfGamesPerTeam  
236                  60             Yes                        468  
4731                 84             Yes                        761  
8240                 73             Yes                        535  
12046                68             Yes                        673  
16677                70             Yes                        804  
19502                76             Yes                        819  
25161                68             Yes                        679  
27997                82             Yes                        625  
33858                81             Yes                        673  
36404                69             Yes                        660  
39858                74             Yes                        765  
44653                91             Yes                        604  
49033                80             Yes                        515  
52600                66             Yes                        461  
57618                75             Yes                        605  
60616                75             Yes                        545  
65999                75             Yes                        644  
70737                59             Yes                        673  
74791                72             Yes                        478  
79181                70              No                        673  
83372                64             Yes                        478  
89542                72             Yes                        652  
94507                74             Yes                        486  
97737                70             Yes                        359  
103723               68             Ye

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [50]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [51]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [52]:
type(df['Wscore'])

pandas.core.series.Series

In [53]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [54]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  TotalNumberOfGamesPerTeam  
0              No                        660  
1              No                        394  
2              No                        746

Here's an equivalent using iloc

In [55]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  TotalNumberOfGamesPerTeam  
0              No                        660  
1              No                        394  
2              No                        746

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [56]:
df.isnull().sum()

Season                       0
Daynum                       0
Wteam                        0
Wscore                       0
Lteam                        0
Lscore                       0
Wloc                         0
Numot                        0
ScoreDifference              0
HighScoringGame              0
TotalNumberOfGamesPerTeam    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y